File: Algorithm/Combination Sum.java
public class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> rst = new ArrayList<List<Integer>>();
        if (candidates == null || candidates.length == 0) {
            return rst;
        }
        List<Integer> items = new ArrayList<Integer>();
        Arrays.sort(candidates);
        helper(rst, items, candidates, target, 0);
        return rst;
    }
    private void helper(List<List<Integer>> rst, List<Integer> items, int[] candidates, int target, int pos) {
        if (target == 0) {
            rst.add(new ArrayList<Integer>(items));
            return;
        } else if (target > 0) {
            for (int i = pos; i < candidates.length; i++) {
                items.add(candidates[i]);
                helper(rst, items, candidates, target - candidates[i], i);
                items.remove(items.size() - 1);
            }
        }
    }
}

File: Algorithm/JavaScript/Combination Sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    let rst = [];

    const helper = (current, sum, index) => {
        if (sum === target) {
            rst.push(current.slice());
            return;
        }
        for (let i = index; i < candidates.length; ++i) {
            if (sum + candidates[i] <= target) {
                helper([...current, candidates[i]], sum + candidates[i], i);
            }
        }
    }

    helper([], 0, 0);
    return rst;
};

console.log(combinationSum([2,3,7], 7));